rag/rag_engine.py

```diff
@@ -0,0 +1,273 @@
+            top_k: Optional[int] = 5,
+            threshold: Optional[float] = None,
+            logging_enabled: Optional[bool] = False,
+    ):
+        """
+        Initialises the RAG engine.
+
+        Args:
+            code_units_path: Path to JSON file containing pre-embedded code units.
+            prompt_model: Name of the large LLM model to use for response
+                generation. Defaults to "gpt-4".
+            embedding_model: Name of the model to use for generating embeddings.
+                Defaults to "microsoft/codebert-base".
+            top_k: Maximum number of similar code units to retrieve. Defaults to 5.
+            threshold: Minimum similarity score (0-1) required for retrieved code
+                units. Defaults to None.
+            logging_enabled: Whether to log detailed information about the RAG
+                pipeline execution. Defaults to False.
+
+        Raises:
+           FileNotFoundError: If the code_units_path file cannot be found.
+           ValueError: If an unsupported prompt_model is specified.
+        """
+        self._prompt_model = prompt_model
+        self._top_k = top_k
+        self._threshold = threshold
+        self._logging_enabled = logging_enabled
+        self._query = ""
+        self._prompt = ""
+        self._response = ""
+        self._similar_code_units = []
+
+        if not os.path.exists(code_units_path):
+            raise FileNotFoundError(
+                f"Embedded code units file not found: {code_units_path}\n" 
+                "Please provide the correct path to the JSON file."
+            )
+
+        # Load code units at initialization
+        with open(code_units_path, "r", encoding="utf-8") as f:
+            self._code_units = json.load(f)
+
+        # Initialize embedder and similarity searcher
+        self._embedder = CodeEmbedder(embedding_model=embedding_model)
+        self._searcher = EmbeddingSimilaritySearch(self._code_units)
+
+        # Initialise large LLM client. At this point, only OpenAI is available.
+        if self._prompt_model in ("gpt-4", "gpt-4o, gpt-4o-mini, gpt-4-turbo"):
+            self._client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
+        else:
+            raise ValueError(f"The model {self._prompt_model} is not supported.")
+
+    def process(self, query: str) -> str:
+        """
+        Processes a query through the complete RAG pipeline.
+
+        This method executes the full retrieval-augmentation-generation sequence:
+        1. Retrieves similar code units based on the query
+        2. Augments the query with the retrieved code context
+        3. Generates a response using the specified LLM
+
+        Args:
+            query: The user's prompt to which RAG will be applied
+
+        Returns:
+            str: The generated response that answers the query using relevant
+                code context.
+        """
+        self._query = query
```

Comment:
Is there a reason for using private variables like `_query` and `_response` instead of just passing them as arguments to the relevant methods and returning them? E.g.,
```suggestion
        similar_code_units = retrieve(query)
        prompt = augment(query, similar_code_units)
        response = generate(prompt)
        
        return response
```

Reply:
just makes it more readable and testable.